src/WCSimConstructRealisticPlacement.cc

```diff
@@ -0,0 +1,1168 @@
+
+    // OD Cap
+    for (int i = -100; i < 100; i++ ){
+      for (int j = -100; j < 100; j++ ){
+
+        G4ThreeVector pmtpos = G4ThreeVector( i*cap_offset_x, j*cap_offset_y, 0.0 );
+
+        if (!(i % 2 == 0)) continue;
+        if (!(j % 3 == 0)) continue;
+
+        bool valid = false;
+        if ((i % 2 == 0 && ((j-3) % 6 == 0) && ((i) % 4 == 0))) valid = true;
+        if ((j % 3 == 0 && ((j) % 6 == 0) && ((i-2) % 4 == 0))) valid = true;
+        if (!valid) continue;
+
+        if (pmtpos.mag() > edgeminimum) continue;
+
+        endcap_assembly_od->AddPlacedVolume(pmtod_dummy_logic, pmtpos, endcapmtprotation);
+      }
+    }
+
+    // Now we can place the end caps, Top first.
+    // -----------------
+    G4ThreeVector topcappos = G4ThreeVector(0.0,0.0,config.InnerDetectorBarrelLength/2);
+    G4RotationMatrix* topcaprot = new G4RotationMatrix();
+    endcap_assembly->MakeImprint( InnerDetectorLogic, topcappos, topcaprot);
+
+    G4ThreeVector topcapposod = G4ThreeVector(0.0,0.0,config.WhiteTyvekBarrelLength/2);
+    G4RotationMatrix* topcaprotod = new G4RotationMatrix();
+    endcap_assembly_od->MakeImprint( OuterDetectorLogic, topcapposod, topcaprotod);
+
+    int pmt20_count_barrel_and_top    = CountLogicalChildren(InnerDetectorLogic, pmt20_dummy_logic);
+    int pmtmulti_count_barrel_and_top = CountLogicalChildren(InnerDetectorLogic, pmtmulti_dummy_logic);
+    int pmtod_count_barrel_and_top    = CountLogicalChildren(OuterDetectorLogic, pmtod_dummy_logic);
+
+    int pmt20_count_top = pmt20_count_barrel_and_top - pmt20_count_barrel;
+    int pmtmulti_count_top = pmtmulti_count_barrel_and_top - pmtmulti_count_barrel;
+    int pmtod_count_top = pmtod_count_barrel_and_top - pmtod_count_barrel;
+
+    // Now cap the oother side
+    // -----------------
+    G4ThreeVector botcappos = G4ThreeVector(0.0,0.0,-config.InnerDetectorBarrelLength/2);
+    G4RotationMatrix* botcaprot = new G4RotationMatrix();
+    botcaprot->rotateX(180*deg);
+    endcap_assembly->MakeImprint( InnerDetectorLogic, botcappos, botcaprot);
+
+    G4ThreeVector botcapposod = G4ThreeVector(0.0,0.0,-config.WhiteTyvekBarrelLength/2);
+    G4RotationMatrix* botcaprotod = new G4RotationMatrix();
+    botcaprotod->rotateX(180*deg);
+    endcap_assembly_od->MakeImprint( OuterDetectorLogic, botcapposod, botcaprotod);
+
+    // Do some final sumation
+    int pmt20_count_total    = CountLogicalChildren(InnerDetectorLogic, pmt20_dummy_logic);
+    int pmtmulti_count_total = CountLogicalChildren(InnerDetectorLogic, pmtmulti_dummy_logic);
+    int pmtod_count_total    = CountLogicalChildren(OuterDetectorLogic, pmtod_dummy_logic);
+
+    int pmt20_count_bottom = pmt20_count_total - pmt20_count_barrel_and_top;
+    int pmtmulti_count_bottom = pmtmulti_count_total - pmtmulti_count_barrel_and_top;
+    int pmtod_count_bottom = pmtod_count_total - pmtod_count_barrel_and_top;
+
+    // We are finished with the assembly, check the totals! 
+    std::cout << "" << std::endl;
```

Comment:
Please use G4cout, G4cerr & G4endl instead
(There was some mangling when mixing these with std::cout etc in the past)